server/src/main/java/org/opensearch/search/pipeline/Pipeline.java

```diff
@@ -117,92 +119,135 @@ protected void afterResponseProcessor(Processor processor, long timeInNanos) {}
-                            logger.warn(
-                                "The exception from request processor ["
-                                    + processor.getType()
-                                    + "] in the search pipeline ["
-                                    + id
-                                    + "] was ignored",
-                                e
-                            );
-                        } else {
-                            throw e;
-                        }
-                    } finally {
-                        long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - start);
-                        afterRequestProcessor(processor, took);
-                    }
+    void transformRequest(SearchRequest request, ActionListener<SearchRequest> requestListener) throws SearchPipelineProcessingException {
+        if (searchRequestProcessors.isEmpty()) {
+            requestListener.onResponse(request);
+            return;
+        }
+
+        try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
+            request.writeTo(bytesStreamOutput);
+            try (StreamInput in = bytesStreamOutput.bytes().streamInput()) {
+                try (StreamInput input = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry)) {
+                    request = new SearchRequest(input);
                 }
-            } catch (Exception e) {
-                onTransformRequestFailure();
-                throw new SearchPipelineProcessingException(e);
-            } finally {
-                long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - pipelineStart);
-                afterTransformRequest(took);
             }
+        } catch (IOException e) {
+            requestListener.onFailure(new SearchPipelineProcessingException(e));
+            return;
+        }
+
+        long[] pipelineStart = new long[1];
+
+        ActionListener<SearchRequest> finalListener = ActionListener.wrap(r -> {
+            long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - pipelineStart[0]);
+            afterTransformRequest(took);
+            requestListener.onResponse(new PipelinedRequest(this, r));
+        }, e -> {
+            long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - pipelineStart[0]);
+            afterTransformRequest(took);
+            onTransformRequestFailure();
+            requestListener.onFailure(new SearchPipelineProcessingException(e));
+        });
+
+        // Chain listeners back-to-front
+        ActionListener<SearchRequest> currentListener = finalListener;
+        for (int i = searchRequestProcessors.size() - 1; i >= 0; i--) {
+            final ActionListener<SearchRequest> nextListener = currentListener;
+            SearchRequestProcessor processor = searchRequestProcessors.get(i);
+            currentListener = ActionListener.wrap(r -> {
```

Comment:
Worked through this logic with pen & paper.

Say you have two request processors P1 & P2.

1) Set up the `finalListener` which takes the last (P2) transformed `SearchRequest` and returns it by wrapping it in a `PipedRequest` object to `TransportSearchAction`.  The chain being set up needs to lead to this.

2) `currentListener` points to `finalListener`.
3) `nextListener` points to what `currentListener` points to, i.e. `finalListener`.
4) `currentListener` sets up the next listener of P2 to point to what `nextListener` points to, i.e. `finalListener`.  Now, P2 correctly executes the final listener.
5) `nextListener` is re-assigned to point to what `currentListener` points to, i.e. P2's listener.
6) `currentListener` sets up the next listener of P1 to point to what `nextListener` points to, i.e. P2's listener.
7) Loop exits.
8) `currentListener` executes, i.e. P1.

Reply:
Correct. (Or at least that's what I was trying to do. It hurt my brain.)

If we didn't have the darned metrics to track (and the `ignoreFailure` flag), the code would be a lot shorter.